lib/internal/streams/readable.js

```diff
@@ -801,7 +801,12 @@ Readable.prototype.pipe = function(dest, pipeOpts) {
   dest.emit('pipe', src);
 
   // Start the flow if it hasn't been started already.
-  if (!state.flowing) {
+
+  if (dest.writableNeedDrain === true) {
+    if (state.flowing) {
+      src.pause();
+    }
```

Comment:
I'm a little unsure how this affects the case where src is already piped to other destinations?

Reply:
:thinking: what about adding the stream to the `state.awaitDrainWriters` (refs `pipeOnDrain` + `ondata(chunk)`). Though that may require copying https://github.com/nodejs/node/pull/35348/files#diff-0117344ddd481d021ad96b9c8eea78a5R741-R747 from `ondata`...

Reply:
I have no idea how that works.

Reply:
Well, same :smile:, based on the initial look it is a list of writables to wait for a drain event so this case seems fitting.
Let's ping @addaleax @BridgeAR @mscdex as most "recent" collaborators according to git-blame.

Reply:
Any chance this can be reverted or done some other way? It seems to cause issues with the source getting stuck in a paused state if you previously fed the output buffer with a lot of data, see eg, https://github.com/electron/asar/issues/210 That module basically just appends a bunch of files to a single binary, but after a random amount of files it gets stuck due to the source being in a paused state.
This could, of course, be easily worked around by adding the writableNeedDrain loop as in the issue above, or adding a resume call to the stream, but I feel like that shouldn't really be necessary.